llvm/tools/sycl-post-link/CompileTimePropertiesPass.cpp

```diff
@@ -102,6 +111,93 @@ Optional<StringRef> getGlobalVariableString(const Value *StringV) {
   return {};
 }
 
+/// Tries to generate a SPIR-V decorate metadata node from an attribute. If
+/// the attribute is unknown \c nullptr will be returned.
+///
+/// @param Ctx   [in] the LLVM context.
+/// @param Attr  [in] the LLVM attribute to generate metadata for.
+///
+/// @returns a pointer to a new metadata node if \c Attr is an attribute with a
+///          known corresponding SPIR-V decorate and the arguments are valid.
+///          Otherwise \c nullptr is returned.
+static MDNode *attributeToDecorateMetadata(LLVMContext &Ctx,
+                                           const Attribute &Attr) {
+  // Currently, only string attributes are supported
+  if (!Attr.isStringAttribute())
+    return nullptr;
+  auto DecorIt = SpirvDecorMap.find(Attr.getKindAsString());
+  if (DecorIt == SpirvDecorMap.end())
+    return nullptr;
+  auto Decor = DecorIt->second;
+  auto DecorCode = Decor.Code;
+  switch (Decor.Type) {
+    case DecorValueTy::uint32:
+      return buildSpirvDecorMetadata(Ctx, DecorCode,
+                                     getAttributeAsInteger<uint32_t>(Attr));
+    case DecorValueTy::boolean:
+      return buildSpirvDecorMetadata(Ctx, DecorCode, hasProperty(Attr));
+    default:
+      llvm_unreachable("Unhandled decorator type.");
+  }
+}
+
+/// Tries to generate a SPIR-V execution mode metadata node from an attribute.
+/// If the attribute is unknown \c None will be returned.
+///
+/// @param Ctx   [in] the LLVM context.
+/// @param Attr  [in] the LLVM attribute to generate metadata for.
+///
+/// @returns a pair with the name of the resulting metadata and a pointer to
+///          the metadata node with its values if the attribute has a
+///          corresponding SPIR-V execution mode. Otherwise \c None is returned.
+static Optional<std::pair<std::string, MDNode *>>
+attributeToExecModeMetadata(LLVMContext &Ctx, const Attribute &Attr) {
+  // Currently, only string attributes are supported
+  if (!Attr.isStringAttribute())
+    return None;
+  StringRef AttrKindStr = Attr.getKindAsString();
+  if (AttrKindStr == "sycl-work-group-size" ||
+      AttrKindStr == "sycl-work-group-size-hint") {
+    // Split values in the comma-separated list integers.
+    SmallVector<StringRef, 3> ValStrs;
+    Attr.getValueAsString().split(ValStrs, ',');
+
+    assert(ValStrs.size() <= 3 &&
+           "sycl-work-group-size and sycl-work-group-size-hint currently only "
+           "support up to three values");
+
+    // SYCL work-group sizes must be reversed for SPIR-V.
+    std::reverse(ValStrs.begin(), ValStrs.end());
+
+    // Get the integers from the strings.
+    auto *SizeTTy = Type::getIntNTy(Ctx, sizeof(size_t) * 8);
```

Comment:
I am a little on the fence. `getLargestLegalIntType` can return 0 if no largest integer type is specified, which it doesn't seem to be for our case. However, the reason we use `size_t` here is because the properties are using host-side `size_t`, so to not lose information it should be that size at least. @Pennycook & @rolandschulz - Is there a reason to use `size_t` for these properties or could we consider making the arguments fixed-sized?

Note that it seems like the SPIR-V translator currently truncs this to 32 bits either way. I am not sure if there is a reason for this.

Reply:
There are some notes about the behavior of `size_t` in SYCL in [Section 5.5](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:scalartypes), clarifying that the host and device must agree on what it means.  I think this means things like the max work-group size and max range size are effectively limited by the host compiler, even if the device is more capable. @gmlueck, do you agree with my reading here?

What does the SPIR-V translator convert this to?  I see two possible [execution modes](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Execution_Mode) defined in the SPIR-V spec:

- `LocalSize`: Indicates the work-group size using 32-bit unsigned integers.
- `LocalSizeId`: Indicates the work-group size using `<id>` operands instead of literals.

Do we need to be using `LocalSizeId` to accept 64-bit unsigned integers?

Reply:
> What does the SPIR-V translator convert this to? I see two possible [execution modes](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Execution_Mode) defined in the SPIR-V spec:
> 
> * `LocalSize`: Indicates the work-group size using 32-bit unsigned integers.
> * `LocalSizeId`: Indicates the work-group size using `<id>` operands instead of literals.
> 
> Do we need to be using `LocalSizeId` to accept 64-bit unsigned integers?

You are right, it does map to `LocalSize`, which in turn expect 32-bit unsigned integers, so the SPIR-V translator is doing the right thing currently. Should we change `work_group_size` and `work_group_size_hint` to use `uint32_t` like `sub_group_size`.

Reply:
Personally I would prefer option 2. As it is with the current version of this, the SPIR-V translator will just clamp it for now, so we can follow-up with changes that allow bigger values through `LocalSizeId` and `LocalSizeHintId` in a future change.

Reply:
> There are some notes about the behavior of size_t in SYCL in [Section 5.5](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:scalartypes), clarifying that the host and device must agree on what it means. I think this means things like the max work-group size and max range size are effectively limited by the host compiler, even if the device is more capable. @gmlueck, do you agree with my reading here?

Did this part of the comment ever get resolved?  I think there are 3 compilers to consider:

(1) The compiler that builds the `sycl-post-link` tool.
(2) The compiler that builds the host part of the user application (aka the "host compiler").
(3) The compiler that builds the device part of the user application (aka the "device compiler").

I agree with @Pennycook's assertion that the size of `size_t` must be the same in both (2) and (3).  However, I think @AlexeySachkov was concerned about compiler (1).  The size of `size_t` in the *source code* of the `sycl-post-link` tool might not be the same as the size of `size_t` in compilers (2) and (3).

Reply:
> The size of size_t in the source code of the sycl-post-link tool might not be the same as the size of size_t in compilers (2) and (3).

I believe you are right. Since sycl-post-link is on the device compiler path, I will have to get the information related to the device, but with the assertion about (2) and (3) having the same `size_t` size, that should be fine. Let's go with that.

Reply:
This was done as part of https://github.com/intel/llvm/pull/7017.